fdbrpc/TokenSign.cpp

```diff
@@ -164,3 +229,193 @@ TEST_CASE("/fdbrpc/TokenSign") {
+		}
+		StringRef key(k, 64);
+		keyPair.privateKeyDer = key;
+		keyPair.publicKeyDer = key;
+		break;
+	}
+	}
+	std::vector<decltype(gen(algo, keyPair.privateKeyDer))> tokens;
+	tokens.reserve(numTokens);
+	fmt::print("Generating tokens");
+	for (int i = 0; i < numTokens; ++i) {
+		tokens.push_back(gen(algo, keyPair.privateKeyDer));
+		if (i % 1000 == 0) {
+			fmt::print(".");
+		}
+	}
+	fmt::print("\n");
+	auto startTime = timer();
+	auto printTime = startTime;
+	double currentTime = startTime;
+	uint64_t iterations = 0;
+	uint64_t lastPrintedIterations = 0;
+	while (startTime + testTime > currentTime) {
+		auto t = tokens[deterministicRandom()->randomInt(0, tokens.size())];
+		validate(algo, t, keyPair.publicKeyDer);
+		if (currentTime - printTime > 1.0) {
+			fmt::print(
+			    "{}s validate {} tokens\n", unsigned(currentTime - startTime), iterations - lastPrintedIterations);
+			lastPrintedIterations = iterations;
+			printTime = currentTime;
+		}
+		currentTime = timer();
+		++iterations;
+	}
+	double endTime = timer();
+	fmt::print("Validated {} tokens in {} seconds ({} tokens per second)\n",
+	           iterations,
+	           endTime - startTime,
+	           unsigned(iterations / (endTime - startTime)));
+}
+
+template <class GenFun, class ValidateFun>
+void tokenVerifyBench(GenFun& gen, ValidateFun& validate) {
+	fmt::print("ES256:\n");
+	fmt::print("======\n");
+	tokenVerifyBench(CryptAlgo::ES256, gen, validate);
+	fmt::print("\n");
+
+	fmt::print("HMAC-SHA256:\n");
+	fmt::print("============\n");
+	tokenVerifyBench(CryptAlgo::HMACSHA256, gen, validate);
+	fmt::print("\n");
+}
+
+Standalone<StringRef> createJWT(CryptAlgo algo,
+                                VectorRef<StringRef> tenants,
+                                double exp,
+                                StringRef keyName,
+                                StringRef key) {
+	rapidjson::StringBuffer headerSS, payloadSS;
+	rapidjson::Writer<rapidjson::StringBuffer> headerW(headerSS), payloadW(payloadSS);
+	headerW.StartObject();
+
+	headerW.String("typ");
+	headerW.String("JWT");
+
+	headerW.String("alg");
+	headerW.String("ES256");
+
+	headerW.EndObject();
+
+	payloadW.StartObject();
+
+	payloadW.String("exp");
+	payloadW.Double(exp);
+
+	payloadW.String("keyName");
+	payloadW.String(reinterpret_cast<const char*>(keyName.begin()), keyName.size());
+
+	payloadW.String("tenants");
+	payloadW.StartArray();
+	for (auto tenant : tenants) {
+		payloadW.String(reinterpret_cast<const char*>(tenant.begin()), tenant.size());
+	}
+	payloadW.EndArray();
+
+	payloadW.EndObject();
+
+	std::string header = headerSS.GetString(), payload = payloadSS.GetString();
+
+	auto unsignedToken = fmt::format("{}.{}", encode64(header), encode64(payload));
+
+	Arena arena;
+	auto signature = sign(algo, arena, StringRef(unsignedToken), key).toString();
+	return Standalone<StringRef>(fmt::format("{}.{}.{}", encode64(header), encode64(payload), encode64(signature)));
```

Comment:
Can reuse the unsigned Token from before without re-computing the base64 encoding.